File: src/fass.js
import fassLexer from "./fassLexer.js";
import fassParser from "./fassParser.js";
import FassListener from "./fassListener.js";
import antlr, { InputStream, CommonTokenStream } from "antlr4";
import { FassErrorListener } from "./error.js";

export function runFass(source) {
	var chars = new InputStream(source, true);
	var lexer = new fassLexer(chars);
	var tokens = new CommonTokenStream(lexer);
	var parser = new fassParser(tokens);
	parser.buildParseTrees = true;
	parser.removeErrorListeners();
	parser.addErrorListener(new FassErrorListener());

	var tree = parser.program();
	var fass = new FassListener();
	antlr.tree.ParseTreeWalker.DEFAULT.walk(fass, tree);

	return {
		constants: fass.constants,
		labels: fass.labels,
		output: fass.output
	};
}

File: src/fass.dart
import "fassLexer.dart";
import "fassParser.dart";
import "MyFassVisitor.dart";
import "package:antlr4/antlr4.dart";

class FassErrorListener extends BaseErrorListener {
  @override
  void syntaxError(
    Recognizer<ATNSimulator> recognizer,
    Object? offendingSymbol,
    int? line,
    int charPositionInLine,
    String msg,
    RecognitionException? e,
  ) {
    final symbol = (offendingSymbol as CommonToken).text;
    throw Exception(
        "Parsing error at line $line:$charPositionInLine. Erroneous symbol: $symbol. «$msg»");
  }
}

MyFassVisitor compile(String code) {
  fassLexer lexer = fassLexer(InputStream.fromString(code));
  lexer.removeErrorListeners();
  lexer.addErrorListener(FassErrorListener());

  CommonTokenStream tokens = CommonTokenStream(lexer);
  fassParser parser = fassParser(tokens);
  parser.removeErrorListeners();
  parser.addErrorListener(FassErrorListener());

  ParseTree tree = parser.program();
  var fass = MyFassVisitor();
  fass.visit(tree);
  return fass;
}
